code/modules/explorer_drone/scanner_array.dm

```diff
@@ -0,0 +1,373 @@
+	var/list/tracked_dishes = list()
+	/// Scan currently in progress if any.
+	var/datum/exoscan/current_scan
+	/// Band for the next wide scan. Increased after successful completion of wide scan.
+	var/wide_scan_band = 1
+	/// Current scan power keyed by site
+	var/list/scan_power_cache = list()
+
+/datum/scanner_controller/proc/create_scan(scan_type,datum/exploration_site/target)
+	if(current_scan)
+		return
+	if(length(GLOB.exoscanner_controller.tracked_dishes) <= 0 || (target && GLOB.exoscanner_controller.get_scan_power(target) <= 0))
+		return
+	current_scan = new(scan_type,target)
+	RegisterSignal(current_scan,COMSIG_PARENT_QDELETING,.proc/cleanup_current_scan)
+	SEND_SIGNAL(src,COMSIG_EXOSCAN_STARTED,current_scan)
+	return current_scan
+
+/datum/scanner_controller/proc/cleanup_current_scan()
+	current_scan = null
+	SEND_SIGNAL(src,COMSIG_EXOSCAN_FINISHED,current_scan)
+
+/datum/scanner_controller/proc/activate_scanner(obj/machinery/exoscanner/scanner)
+	if(scanner in tracked_dishes)
+		return
+	tracked_dishes += scanner
+	update_scan_power()
+
+/datum/scanner_controller/proc/deactivate_scanner(obj/machinery/exoscanner/scanner)
+	if(!(scanner in tracked_dishes))
+		return
+	tracked_dishes -= scanner
+	update_scan_power()
+
+/datum/scanner_controller/proc/update_scan_power()
+	scan_power_cache = list()
+	if(current_scan) //Check if we need to interrupt current scan.
+		var/current_power = length(tracked_dishes)
+		if(current_scan.target)
+			current_power = get_scan_power(current_scan.target)
+		if(current_scan.scan_power > current_power)
+			current_scan.stop("Scan swarm power reduced")
+
+/datum/scanner_controller/proc/get_scan_power(datum/exploration_site/target)
+	if(!scan_power_cache[target])
+		scan_power_cache[target] = calculate_scan_power(target.scan_conditions)
+	return scan_power_cache[target]
+
+/datum/scanner_controller/proc/calculate_scan_power(conditions)
+	. = 0
+	for(var/obj/machinery/exoscanner/dish in tracked_dishes)
+		var/effective_power = 1
+		for(var/datum/scan_condition/condition in conditions)
+			effective_power *= condition.check_dish(dish)
+			if(!effective_power) //Don't bother continuing if it's zero
+				break
+		. += effective_power
+
+/// Scan condition, these require some specific setup for the dish to count for the scan power for the given site
+/datum/scan_condition
+	var/name
+	var/description
+
+/// Returns power multiplier of the dish depending on condition.
+/datum/scan_condition/proc/check_dish(obj/machinery/exoscanner/dish)
+	return 1
+
+/datum/scan_condition/nebula
+	name = "Nebula"
+	description = "Site is within a unusually dense nebula, to reduce scanner noise position dishes at least 15 tiles apart"
+	var/distance = 15
+
+/datum/scan_condition/nebula/check_dish(obj/machinery/exoscanner/dish)
+	for(var/obj/machinery/exoscanner/other_dish in GLOB.exoscanner_controller.tracked_dishes)
+		if(dish != other_dish && dish.z == other_dish.z && get_dist(dish,other_dish) < distance)
```

Comment:
Won't this work just as fine?
```suggestion
	for(var/obj/machinery/exoscanner/other_dish in GLOB.exoscanner_controller.tracked_dishes - dish)
		if(dish.z == other_dish.z && get_dist(dish,other_dish) < distance)
```

Reply:
![image](https://user-images.githubusercontent.com/4047233/111916475-5cf36e80-8a7b-11eb-9f17-4c48eb3e0176.png)
This is with a list of size 10.